server/src/core/invitation/invitation-service.ts

```diff
@@ -0,0 +1,131 @@
+import { Db, Collection } from 'mongodb';
+import { generateId } from '../util';
+import { Invitation, DbInvitation, InvitationCreateArgs, InvitationService, InvitationStatus } from './types';
+import { createDbOpFailedError, rethrowIfAppError, createResourceNotFoundError, AppError } from '../error';
+import * as validators from './validator';
+import * as messages from '../messages';
+
+const COLLECTION = 'invitations';
+
+export class Invitations implements InvitationService {
+  private db: Db;
+  private collection: Collection<DbInvitation>;
+  private indexesCreated: boolean;
+
+  constructor(db: Db) {
+    this.db = db;
+    this.collection = this.db.collection(COLLECTION);
+    this.indexesCreated = false;
+  }
+
+  async createIndexes(): Promise<void> {
+    if (this.indexesCreated) return;
+
+    try {
+      /*
+      Expire every invitation document 
+      after 172800 seconds (i.e., two days) have passed since creation
+      */
+      await this.collection.createIndex(
+        { expiresAt: 1},
+        { expireAfterSeconds: 172800 }
+      );
+      
+      this.indexesCreated = true;
+    }
+    catch (e) {
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async create(args: InvitationCreateArgs): Promise<Invitation> {
+    const { invitorId, invitorName, inviteeName, inviteePhone, inviteeEmail, inviteeRole, hasAccount } = args;
+    const now = new Date();
+    const invitation: DbInvitation = {
+      _id: generateId(),
+      invitorId, 
+      invitorName, 
+      inviteeName,
+      inviteePhone,
+      inviteeEmail,
+      inviteeRole,
+      hasAccount,
+      status: 'pending',
+      expiresAt: now,
+      createdAt: now,
+      updatedAt: now,
+    }
+
+    try {
+      const res = await this.collection.insertOne(invitation);
+      return res.ops[0];
+    }
+    catch (e) {
+      rethrowIfAppError(e);
+
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async accept(invitationId: string): Promise<Invitation> {
+    try {
+      const result = await this.collection.findOneAndUpdate(
+        { _id: invitationId },
+        { $set: { status: 'accepted' } },
+        { upsert: true, returnOriginal: false }
+      );
+      
+      if (!result) throw createResourceNotFoundError(messages.ERROR_INVITATION_NOT_FOUND);
+      
+      return result.value;
+    }
+    catch(e) {
+      rethrowIfAppError(e);
+
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async reject(invitationId: string): Promise<Invitation> {
+    try {
+      const result = await this.collection.findOneAndUpdate(
+        { _id: invitationId },
+        { $set: { status: 'rejected' } },
+        { upsert: true, returnOriginal: false }
+      );
```

Comment:
You should only be able to reject an existing rejection. See the notes on accepting invitations.

Actually these two methods are similar, the only difference is the `status`. You can create a private method that encapsulates this logic for updating the status and then call that private method from both `accept()` and `reject()`.